Buy/Graph.Task.swift

```diff
@@ -26,42 +26,271 @@
+        internal let cachePolicy:  CachePolicy
+        internal let retryHandler: RetryHandler<R>?
+        internal let completion:   TaskCompletion
+        
+        internal var task:         URLSessionDataTask?
         
         // ----------------------------------
         //  MARK: - Init -
         //
-        internal init(representing task: URLSessionDataTask) {
-            self.task = task
+        internal init(session: URLSession, cache: Cache, request: URLRequest, cachePolicy:  CachePolicy, retryHandler: RetryHandler<R>? = nil, completion: @escaping TaskCompletion) {
+            self.cache        = cache
+            self.session      = session
+            self.request      = request
+            self.cachePolicy  = retryHandler == nil ? cachePolicy : .networkOnly
+            self.retryHandler = retryHandler
+            self.completion   = completion
         }
         
         // ----------------------------------
-        //  MARK: - Setters -
+        //  MARK: - Control -
         //
-        internal func setTask(_ task: URLSessionDataTask) {
-            self.task = task
+        func resume() {
+            self.isResumed = true
+            self.resume(using: self.cachePolicy)
+        }
+        
+        private func resume(using cachePolicy: CachePolicy) {
+            
+            let request    = self.request
+            let completion = self.completion
+            let hash       = request.hash
+            
+            switch cachePolicy {
+            case .cacheFirst(let expireIn):
+                print("Exercising cache policy: CACHE_FIRST(\(expireIn))")
+                
+                self.cachedModelFor(hash, expireIn: expireIn) { response in
+                    if let response = response {
+                        completion(response, nil)
+                        return
+                    } else {
+                        self.resume(using: .networkOnly)
+                    }
+                }
+                
+            case .cacheOnly:
+                print("Exercising cache policy: CACHE_ONLY")
+                
+                self.cachedModelFor(hash) { response in
+                    completion(response, nil)
+                    return
+                }
+                
+            case .networkFirst(let expireIn):
+                print("Exercising cache policy: NETWORK_FIRST(\(expireIn))")
+                
+                self.task = self.graphTaskWith(request, retryHandler: self.retryHandler) { response, data, error in
+                    
+                    if let _ = response, let data = data {
+                        self.cache(data, for: hash)
+                        completion(response, error)
+                        
+                    } else {
+                        
+                        self.cachedModelFor(hash, expireIn: expireIn) { response in
+                            if let response = response {
```

Comment:
Because we have hit the cache only after a network error, which means the `error` is always not `nil` in this scope. We don't want to return the error if the cache has a response.